File: helpers/plot/plot.go
package plot

import (
	"encoding/json"
	"fmt"
	"sort"
	"time"

	"github.com/open-traffic-generator/conformance/helpers/table"
)

type DurRange []int64

func (a DurRange) Len() int           { return len(a) }
func (a DurRange) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a DurRange) Less(i, j int) bool { return a[i] < a[j] }

type Distribution struct {
	ApiName  string        `json:"api_name"`
	Duration time.Duration `json:"duration"`
	Type     string
}

type PerApiDistribution struct {
	ApiName         string `json:"api_name"`
	sortedDurations DurRange
	Max             int64 `json:"max"`
	Min             int64 `json:"min"`
	Avg             int64 `json:"avg"`
	P50             int64 `json:"p50"`
	P75             int64 `json:"p75"`
	P90             int64 `json:"p90"`
	P95             int64 `json:"p95"`
	P99             int64 `json:"p99"`
}

type Plot struct {
	Name            string               `json:"name"`
	Iterations      int                  `json:"iterations"`
	Durations       []Duration           `json:"durations"`
	Distributions   []PerApiDistribution `json:"distributions"`
	orderedApiNames []string
}

func NewPlot() *Plot {
	return &Plot{
		Durations: make([]Duration, 0),
	}
}

func (p *Plot) AppendDuration(d Duration) {
	p.Durations = append(p.Durations, d)
}

func (p *Plot) AppendZero() {
	p.Durations = append(p.Durations, Duration{Duration: 0})
}

func (p *Plot) CalculateIterations() {
	p.Iterations = 0

	for i := range p.Durations {
		if p.Durations[i].ApiName == "" {
			p.Iterations += 1
		}
	}
}

func (p *Plot) ApiDurationsMap() map[string]DurRange {
	apiDursMap := map[string]DurRange{}
	p.orderedApiNames = []string{}

	for _, d := range p.Durations {
		if d.ApiName == "" {
			continue
		}

		if _, ok := apiDursMap[d.ApiName]; !ok {
			apiDursMap[d.ApiName] = DurRange{}
			p.orderedApiNames = append(p.orderedApiNames, d.ApiName)
		}

		apiDursMap[d.ApiName] = append(apiDursMap[d.ApiName], int64(d.Duration))
	}

	return apiDursMap
}

func (p *Plot) GetPercentileDuration(durs DurRange, percent int) int64 {
	return durs[(percent*len(durs))/100]
}

func (p *Plot) CalcDistributions() {
	apiDursMap := p.ApiDurationsMap()

	p.Distributions = make([]PerApiDistribution, 0, len(apiDursMap))
	for _, name := range p.orderedApiNames {
		durs := apiDursMap[name]

		sort.Sort(durs)
		d := PerApiDistribution{
			ApiName:         name,
			sortedDurations: durs,
			Min:             durs[0],
			Max:             durs[len(durs)-1],
			Avg: func() int64 {
				sum := int64(0)
				for _, dur := range durs {
					sum += dur
				}

				return sum / int64(len(durs))
			}(),
			P50: p.GetPercentileDuration(durs, 50),
			P75: p.GetPercentileDuration(durs, 75),
			P90: p.GetPercentileDuration(durs, 90),
			P95: p.GetPercentileDuration(durs, 95),
			P99: p.GetPercentileDuration(durs, 99),
		}

		p.Distributions = append(p.Distributions, d)
	}
}

func (p *Plot) Analyze(name string) {
	p.Name = name
	p.CalculateIterations()
	p.CalcDistributions()
}

func (p *Plot) ToJson() (string, error) {
	b, err := json.MarshalIndent(p, "", "  ")
	if err != nil {
		return "", err
	}

	return string(b), nil
}

func (p *Plot) ToTable() (string, error) {
	tb := table.NewTable(
		fmt.Sprintf("Distribution: %s (Iterations %d)", p.Name, p.Iterations),
		func() []string {
			cols := make([]string, 0, len(p.Distributions)+1)
			cols = append(cols, "Dist")
			for _, d := range p.Distributions {
				cols = append(cols, d.ApiName)
			}

			return cols
		}(),
		20,
	)

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "min")
		for _, d := range p.Distributions {
			row = append(row, d.Min)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "avg")
		for _, d := range p.Distributions {
			row = append(row, d.Avg)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "max")
		for _, d := range p.Distributions {
			row = append(row, d.Max)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "p50")
		for _, d := range p.Distributions {
			row = append(row, d.P50)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "p75")
		for _, d := range p.Distributions {
			row = append(row, d.P75)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "p90")
		for _, d := range p.Distributions {
			row = append(row, d.P90)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "p95")
		for _, d := range p.Distributions {
			row = append(row, d.P95)
		}
		return row
	})())

	tb.AppendRow((func() []interface{} {
		row := make([]interface{}, 0, len(p.Distributions)+1)
		row = append(row, "p99")
		for _, d := range p.Distributions {
			row = append(row, d.P99)
		}
		return row
	})())

	return tb.String(), nil
}

File: helpers/plot/plot.py
import datetime
import json

from helpers.table import table


class Distribution(object):
    def __init__(self, api_name, duration, typ):
        self.api_name = api_name
        self.duration = duration
        self.typ = typ


class PerApiDistribution(object):
    def __init__(self, api_name):
        self.api_name = api_name
        self.sorted_durations = []
        self.max = 0
        self.min = 0
        self.avg = 0
        self.p50 = 0
        self.p75 = 0
        self.p90 = 0
        self.p95 = 0
        self.p99 = 0


class Duration(object):
    def __init__(self, api_name, duration, time):
        self.api_name = api_name
        self.duration = duration
        self.time = time


class Plot(object):
    def __init__(self):
        self.name = ""
        self.iterations = 0
        self.durations = []
        self.distributions = []
        self.ordered_api_names = []

    def append_duration(self, duration):
        if duration is not None:
            self.durations.append(duration)

    def append_zero(self):
        self.durations.append(Duration("", 0, None))

    def calculate_iterations(self):
        self.iterations = 0

        for d in self.durations:
            if d.api_name == "":
                self.iterations += 1

    def api_duration_map(self):
        api_dur_map = {}
        self.ordered_api_names = []

        for d in self.durations:
            if d.api_name == "":
                continue

            if d.api_name not in api_dur_map:
                api_dur_map[d.api_name] = []
                self.ordered_api_names.append(d.api_name)

            api_dur_map[d.api_name].append(d.duration)

        return api_dur_map

    def get_percentile_duration(self, durations, percent):
        return durations[int((percent * len(durations)) / 100)]

    def calc_distributions(self):
        api_dur_map = self.api_duration_map()
        self.distributions = []

        for name in self.ordered_api_names:
            durations = sorted(api_dur_map[name])

            d = PerApiDistribution(name)
            d.sorted_durations = durations
            d.min = durations[0]
            d.max = durations[len(durations) - 1]
            d.avg = int(sum(durations) / len(durations))
            d.p50 = self.get_percentile_duration(durations, 50)
            d.p75 = self.get_percentile_duration(durations, 75)
            d.p90 = self.get_percentile_duration(durations, 90)
            d.p95 = self.get_percentile_duration(durations, 95)
            d.p99 = self.get_percentile_duration(durations, 99)

            self.distributions.append(d)

    def analyze(self, name):
        self.name = name
        self.calculate_iterations()
        self.calc_distributions()

    def to_json(self):
        return json.dumps(
            self,
            default=lambda o: o.microsecond * 1000
            if isinstance(o, datetime.datetime)
            else o.__dict__,
            indent="  ",
        )

    def to_table(self):
        tb = table.Table(
            "Distribution: %s (Iterations %d)" % (self.name, self.iterations),
            ["Dist"] + [d.api_name for d in self.distributions],
            col_width=25,
        )

        tb.append_row(["min"] + [d.min for d in self.distributions])
        tb.append_row(["avg"] + [d.avg for d in self.distributions])
        tb.append_row(["max"] + [d.max for d in self.distributions])
        tb.append_row(["p50"] + [d.p50 for d in self.distributions])
        tb.append_row(["p75"] + [d.p75 for d in self.distributions])
        tb.append_row(["p90"] + [d.p90 for d in self.distributions])
        tb.append_row(["p95"] + [d.p95 for d in self.distributions])
        tb.append_row(["p99"] + [d.p99 for d in self.distributions])

        return str(tb)
